### PR TITLE
Introduce creation of parameterized tooling models

### PR DESCRIPTION
**Pull request by lsmaira**
This commit introduces the possibility of passing parameters through the
Tooling API to the model builders in order to create models based on
information received by the client.

This feature allows plugins to register a model builder for a given
model and parameter and then build models based on the received
parameter. It is preferable to passing gradle properties for two
reasons: first convenience and second because parameters can be created
inside the BuildAction.

A new interface ToolingParameterizedModelBuilder was created and should
be extended by parameterized builders. New methods were added to the
BuildController in order to build models with parameters. In order to
keep cross version compatibility, new protocol interfaces were created.

### Context

This change will be useful for Android Studio to sync only one variant of Android projects instead of all of them.

Problem and solution discussed with Gradlers in private meetings instead of public forums.

Design doc included.

As suggested in Slack, a normal Java Proxy was used in client side for the moment instead of refactoring the managed proxies.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [x] [Link to Design Spec](https://github.com/gradle/gradle/tree/master/design-docs) for changes that affect more than 1 public API (that is, not in an `internal` package) or updates to > 20 files
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by oehme** [comment]
Thank you for the PR Lucas, we will prioritise reviewing this. Let's target Gradle 4.3 (which comes out before AS 3.1) for this change. If we make it into 4.2, all the better. But a change of this size will likely take longer to review.

**Comment by lsmaira** [comment]
Thank you! I'll send another commit later changing versions in javadocs then

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import org.gradle.tooling.internal.consumer.versioning.VersionDetails;
+import org.gradle.tooling.internal.protocol.BuildResult;
+import org.gradle.tooling.internal.protocol.ConnectionVersion4;
+import org.gradle.tooling.internal.protocol.InternalBuildActionFailureException;
+import org.gradle.tooling.internal.protocol.InternalBuildCancelledException;
+import org.gradle.tooling.internal.protocol.InternalCancellableConnectionVersion2;
+
+import java.io.File;
+
+public class ParameterCancellableConsumerConnection extends TestExecutionConsumerConnection {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java]: There seems to be quite a bit of duplicated code between this and CancellableConsumer. Can you please refactor this?

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java [hunk 2]
```diff
@@ -128,6 +129,16 @@ public Object run(InternalBuildAction<?> clientAction, BuildCancellationToken ca
         return run(action, cancellationToken, listenerConfig, providerParameters, params);
     }
 
+    public Object run(InternalBuildActionVersion2<?> clientAction, BuildCancellationToken cancellationToken, ProviderOperationParameters providerParameters) {
```
Comment (oehme) [inline subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java]: There's quite a bit of duplicated code here, can you refactor this?

File: subprojects/core/src/main/java/org/gradle/tooling/provider/model/ToolingParameterizedModelBuilder.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * <p>Although it is not enforced, the model object should be immutable, as the tooling API will do some caching and other performance optimizations on the
+ * assumption that the model is effectively immutable. The tooling API does not make any guarantees about how the client application will use the model object.</p>
+ *
+ * @param <T> The type of parameter used by this model builder.
+ * @since 4.2
+ * @see ToolingModelBuilder
+ */
+@Incubating
+public interface ToolingParameterizedModelBuilder<T> extends ToolingModelBuilder {
```
[the remark this answers is not in the document]
Comment (lsmaira) [reply subprojects/core/src/main/java/org/gradle/tooling/provider/model/ToolingParameterizedModelBuilder.java]: Done.

File: design-docs/tooling-api-parameterized-custom-models.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
+
+# Feature: Gradle plugin builds custom tooling models using parameter passed by client
```
Comment (oehme) [inline design-docs/tooling-api-parameterized-custom-models.md]: You can remove this doc, we already have the Google Doc and we recently deleted the design-docs folder, because they just go out of date over time anyway :)

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import org.gradle.tooling.internal.protocol.InternalCancellableConnectionVersion2;
+import org.gradle.tooling.internal.protocol.InternalCancellationToken;
+import org.gradle.tooling.internal.protocol.ModelIdentifier;
+
+/**
+ * An adapter for {@link InternalCancellableConnectionVersion2}.
+ *
+ * <p>Used for providers >= 4.3.</p>
+ */
+public class ParameterCancellableConsumerConnection extends TestExecutionConsumerConnection {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java]: Cancellation is not new and so shouldn't be in the name. It should be something like `ParameterAcceptingConsumerConnection`

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalCancellableConnectionVersion2.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * <p>DO NOT CHANGE THIS INTERFACE - it is part of the cross-version protocol.
+ *
+ * <p>Consumer compatibility: This interface is used by all consumer versions from 4.3.</p>
+ * <p>Provider compatibility: This interface is implemented by all provider versions from 4.3.</p>
+ *
+ * @since 4.3
+ * @see org.gradle.tooling.internal.protocol.ConnectionVersion4
+ */
+public interface InternalCancellableConnectionVersion2 extends InternalProtocolInterface {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalCancellableConnectionVersion2.java]: This is a ParameterAcceptingConnection. Cancellation is not new here.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalCancellableConnectionVersion2.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+     * @param modelIdentifier The identifier of the model to build.
+     * @param cancellationToken The token to propagate cancellation.
+     * @throws BuildExceptionVersion1 On build failure.
+     * @throws InternalUnsupportedModelException When the requested model is not supported.
+     * @throws InternalUnsupportedBuildArgumentException When the specified command-line options are not supported.
+     * @throws InternalBuildCancelledException When the operation was cancelled before it could complete.
+     * @throws IllegalStateException When this connection has been stopped.
+     * @since 4.3
+     */
+    BuildResult<?> getModel(ModelIdentifier modelIdentifier, InternalCancellationToken cancellationToken,
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalCancellableConnectionVersion2.java]: This method is unrelated to parameterized requests and shouldn't be here.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    private final ActionRunner actionRunner;
+    private final ModelProducer modelProducer;
+
+    public ParameterCancellableConsumerConnection(ConnectionVersion4 delegate, ModelMapping modelMapping, ProtocolToModelAdapter adapter) {
+        super(delegate, modelMapping, adapter);
+        final InternalCancellableConnectionVersion2 connection = (InternalCancellableConnectionVersion2) delegate;
+        InternalCancellableConnectionWrapper connectionWrapper = new InternalCancellableConnectionWrapper() {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java]: This wrapper should not be necessary. The old getModel method is unchanged and the new run method is only available here.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/CancellableConsumerConnection.java [hunk 9]
```diff
@@ -78,12 +92,12 @@ public RuntimeException transform(RuntimeException e) {
         }
     }
 
-    private static class CancellableActionRunner implements ActionRunner {
-        private final InternalCancellableConnection executor;
+    protected static class CancellableActionRunner implements ActionRunner {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/CancellableConsumerConnection.java]: Should be in its own file instead of protected static nested class.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/CancellableConsumerConnection.java [hunk 10]
```diff
@@ -66,7 +80,7 @@ protected ModelProducer getModelProducer() {
         return modelProducer;
     }
 
-    private static class ExceptionTransformer implements Transformer<RuntimeException, RuntimeException> {
+    protected static class ExceptionTransformer implements Transformer<RuntimeException, RuntimeException> {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/CancellableConsumerConnection.java]: Should be in its own class instead of protected static.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/CancellableConsumerConnection.java [hunk 11]
```diff
@@ -106,4 +120,12 @@ private CancellableActionRunner(InternalCancellableConnection executor, Transfor
             return result.getModel();
         }
     }
+    /**
+     * Interface representing either an {@link InternalCancellableConnection}
+     * or an {@link org.gradle.tooling.internal.protocol.InternalCancellableConnectionVersion2}.
+     */
+    protected interface InternalCancellableConnectionWrapper {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/CancellableConsumerConnection.java]: Should not be necessary. Instead, the new connection should use a subclass of CancellableActionRunner which then also supports the new parameterized actions.

File: design-docs/tooling-api-parameterized-custom-models.md (hunk 4 above)
Comment (lsmaira) [reply design-docs/tooling-api-parameterized-custom-models.md]: Cool! Done.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java (hunk 5 above)
Comment (lsmaira) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/ParameterCancellableConsumerConnection.java]: Done.

**Comment by oehme** [comment]
Thanks for letting me know Lucas. I'll merge the change into master then instead, which means we could do additional changes without another protocol change.

**Comment by lsmaira** [comment]
Just rebased and changed version to 4.4 in all javadocs, comments, etc.

**Comment by oehme** [comment]
Thank you again for the excellent work Lucas!

**Comment by lsmaira** [comment]
Thank YOU for the help and reviews :)
